File: htmlapp/frontend/src/engine/services/package_registry.ts
/**
 * Package registry — in-memory index built from IndexedDB package records.
 *
 * Ported from `app/api/aica_api/services/package_registry.py` (behavior-of-record).
 * The Python registry scans a `packages_dir` for `package.json` files at
 * construction time. The htmlapp offline build has no filesystem: package
 * manifests are bundled JSON, generated into the data registry
 * (`../../data/registry.ts`) and seeded into IndexedDB by
 * `storage/db.ts#seedDefaults()` on first launch. This registry reads the
 * already-seeded `PackageRecord` rows via `packages_store` instead of
 * scanning a directory.
 *
 * `listSummaries()`'s `errors` surfaces
 * `../../data/packages#builtinPackageErrors()` — one entry per bundled
 * manifest that fails the required-field validation in
 * `../../data/packages/validate.ts` (mirrors Python's `package.py` Pydantic
 * model, including its non-empty `compatible_scenario_types` requirement).
 * `seedDefaults()` never seeds an invalid manifest into IndexedDB in the
 * first place — it only ever calls `builtinPackages()`, which already
 * excludes them — so `errors` here answers "why isn't a bundled package I
 * expected available," the same question Python's `list_errors()` answers
 * for manifests that failed Pydantic validation during its `packages_dir`
 * scan, not "something already seeded is broken."
 *
 * Public API mirrors the Python registry:
 *   packageRegistry.listSummaries() -> Promise<{ packages, errors }>
 *   packageRegistry.get(id)         -> Promise<PackageManifest>
 *   packageRegistry.isCompatible(package, scenario) -> boolean
 */
import { packagesStore } from '../../storage/packages_store'
import { builtinPackageErrors } from '../../data/packages'
import type { PackageManifest, PackageSummary, RegistryError, ScenarioDef } from '../../api/types'

export const packageRegistry = {
  /** Return lightweight summaries for all seeded packages, plus any bundled-manifest load errors. */
  async listSummaries(): Promise<{ packages: PackageSummary[]; errors: RegistryError[] }> {
    const records = await packagesStore.list()
    const packages: PackageSummary[] = records.map((rec) => ({
      id: rec.manifest.id,
      version: rec.manifest.version,
      label: rec.manifest.label,
      algorithm_type: rec.manifest.algorithm.type,
      compatible_scenario_types: rec.manifest.compatible_scenario_types,
    }))
    return { packages, errors: builtinPackageErrors() }
  },

  /** Return the full PackageManifest for a package_id. Throws if not found. */
  async get(id: string): Promise<PackageManifest> {
    const rec = await packagesStore.get(id)
    if (!rec) {
      throw new Error(`Package '${id}' not found or invalid`)
    }
    return rec.manifest
  },

  /**
   * Return true if the scenario type is in the package's compat list.
   * Defensive: a manifest with a missing or non-array
   * `compatible_scenario_types` is treated as compatible with nothing,
   * rather than throwing. `builtinPackages()` already excludes bundled
   * manifests missing the field, so this is belt-and-braces there, but this
   * is also reachable outside React (e.g. a worker `dispatch()` call) where
   * nothing upstream guards the field the way `ScenarioSelector.tsx`
   * assumes — without this guard, reaching it produces a raw TypeError
   * instead of a domain-shaped `false`.
   */
  isCompatible(pkg: PackageManifest, scenario: ScenarioDef): boolean {
    const types = pkg.compatible_scenario_types
    return Array.isArray(types) && types.includes(scenario.type)
  },
}

File: app/api/aica_api/services/package_registry.py
"""Package registry (T019) — scans packages_dir and loads PackageManifests.

Scans the packages directory for subdirectories each containing a
``package.json`` file.  Valid manifests are loaded and indexed by id;
invalid ones are reported in an errors list and never partially used.

Public API:
  PackageRegistry(packages_dir: Path)
    .list_summaries() -> list[dict]   — id, version, label, compatible_scenario_types
    .list_errors()    -> list[dict]   — {package_dir, error} for invalid packages
    .get(package_id)  -> PackageManifest | None
    .is_compatible(package, scenario) -> bool
"""

from __future__ import annotations

import pathlib
from typing import Any

from pydantic import ValidationError

from aica_api.models.package import PackageManifest
from aica_api.models.scenario import ScenarioDef
from aica_api.storage.file_store import read_json


class PackageRegistry:
    """In-memory index of all available packages in a packages directory."""

    def __init__(self, packages_dir: pathlib.Path) -> None:
        self._packages: dict[str, PackageManifest] = {}
        self._errors: list[dict[str, Any]] = []
        self._scan(packages_dir)

    # ── Public API ─────────────────────────────────────────────────────────

    def list_summaries(self) -> list[dict]:
        """Return lightweight summaries for all successfully loaded packages."""
        return [
            {
                "id": pkg.id,
                "version": pkg.version,
                "label": pkg.label,
                "compatible_scenario_types": pkg.compatible_scenario_types,
                "algorithm_type": pkg.algorithm.type,
            }
            for pkg in self._packages.values()
        ]

    def list_errors(self) -> list[dict]:
        """Return error entries for packages that failed to load."""
        return list(self._errors)

    def get(self, package_id: str) -> PackageManifest | None:
        """Return the full PackageManifest for a package_id, or None."""
        return self._packages.get(package_id)

    def is_compatible(
        self, package: PackageManifest, scenario: ScenarioDef
    ) -> bool:
        """Return True if the scenario type is in the package's compat list."""
        return scenario.type in package.compatible_scenario_types

    # ── Private helpers ────────────────────────────────────────────────────

    def _scan(self, packages_dir: pathlib.Path) -> None:
        """Scan packages_dir for subdirectories containing package.json."""
        if not packages_dir.exists():
            return

        for pkg_dir in sorted(packages_dir.iterdir()):
            if not pkg_dir.is_dir():
                continue
            manifest_path = pkg_dir / "package.json"
            if not manifest_path.exists():
                continue
            try:
                data = read_json(str(manifest_path))
                # Proposal-family packages (they declare `kind`/`family`, e.g. the
                # transparent content selector) are owned by the separate proposal
                # package registry, not the trigger registry. Skip them silently —
                # they are not trigger packages and must not surface as errors here.
                if data.get("kind") is not None or data.get("family") is not None:
                    continue
                manifest = PackageManifest(**data)
                self._packages[manifest.id] = manifest
            except (ValidationError, Exception) as exc:
                self._errors.append(
                    {
                        "package_dir": str(pkg_dir),
                        "error": str(exc),
                    }
                )
